### PR TITLE
[FLINK-34470][Connectors/Kafka] Fix indefinite blocking by adjusting stopping condition in split reader

### PR DESCRIPTION
**Pull request by dongwoo6kim**
## Problem 
When using the flink kafka connector in batch scenarios, consuming transactional messages can cause indefinite hanging.
This issue can be easily reproduced with following steps. 
1. Produce transactional messages and commit them.
2. Configure `scan.bounded.mode` to `latest-offset` and run consumer using flink kafka connector

## Cause
The previous stopping condition in the `KafkaPartitionSplitReader` compared the offset of the last record with the `stoppingOffset`. This approach works for streaming use cases and batch processing of non-transactional messages. However, in scenarios involving transactional messages, this is insufficient.      
[Control messages](https://kafka.apache.org/documentation/#controlbatch), which are not visible to clients, can occupy the entire range between the last record's offset and the stoppingOffset which leads to indefinite blocking.

## Workaround
I've modified the stopping condition to use `consumer.position(tp)`, which effectively skips any control messages present in the current poll, pointing directly to the next record's offset. 
To handle edge cases, particularly when `properties.max.poll.records` is set to `1`, I've adjusted the fetch method to always check all assigned partitions, even if no records are returned in a poll.

#### Edge case example 
Consider partition `0`, where offsets `13` and `14` are valid records and `15` is a control record. If `stoppingOffset` is set to 15 for partition `0`and `properties.max.poll.records` is configured to `1`, checking only partitions that return records would miss offset 15. By consistently reviewing all assigned partitions, the consumer’s position jumps control record in the subsequent poll, allowing the system to escape.

## Discussion
To address the metric issue in [FLINK-33484](https://issues.apache.org/jira/browse/FLINK-33484), I think we need to make wrapper class of  `ConsumerRecord` for example `ConsumerRecordWithOffsetJump`.
```java
public ConsumerRecordWithOffsetJump(ConsumerRecord<K, V> record, long offsetJump) {
        this.record = record;
        this.offsetJump = offsetJump;
    }
```
And we may need new `KafkaPartitionSplitReader` that implements     
`SplitReader<ConsumerRecordWithOffsetJump<byte[], byte[]>, KafkaPartitionSplit>`.     
So when record is emitted it should set current offset not just `record.offset()+1` but     
`record.offset() + record.jumpValue` in [here](https://github.com/apache/flink-connector-kafka/blob/369e7be46a70fd50d68746498aed82105741e7d6/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaRecordEmitter.java#L54).  
  `jumpValue` is typically 1, except for the last record of each poll where it's calculated as     
`consumer.position() - lastRecord.offset()`.       
If this sounds good to everyone, I'm happy to work on this.

### CODE REVIEW

**Comment by LinMingQiang** [comment]
its work , i had try.

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (LinMingQiang) [inline flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:142]: consumerRecords.partitions().forEach(trackTp -> {
                // Track this partition's record lag if it never appears before
                kafkaSourceReaderMetrics.maybeAddRecordsLagMetric(consumer, trackTp);
            });

Comment (LinMingQiang) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:142]: we do not need to track tp when consumerRecords is empty.

Comment (dongwoo6kim) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java]: @LinMingQiang, thanks for the review. 
I've changed to track tp, only when records for that tp is not empty.

**Review by morazow — APPROVED** [review]
Hey @dongwoo6kim, overall seems okay from my side. But would be possible to add a integration-test ITCase for this case

**Comment by dongwoo6kim** [comment]
Hello @morazow I've added ITCase for this case. 
This case fails for current main branch due to timeout and works okay with fixed code.

**Review by morazow — APPROVED** [review]

**Comment by morazow** [comment]
Thanks @dongwoo6kim ,

Tests looks good from my side 👍 

(_Recently I faced similar issue which maybe related, when running batch mode with setting `startingOffsets`. The change should solve that issue. But we may create issue for it_)

**Comment by dongwoo6kim** [comment]
Thanks for confirming @morazow, 
Please feel free to provide any additional advice before merging this fix. 
It would be also helpful if you could elaborate more on the issue you mentioned and consider adding relevant test code for it.

**Comment by morazow** [comment]
Hey @dongwoo6kim, we created another [issue](https://issues.apache.org/jira/browse/FLINK-35565) for it, the solution seems to be similar but let's discuss it again once this PR is merged.

**Comment by dongwoo6kim** [comment]
Hello @morazow, I've added [test code](https://github.com/apache/flink-connector-kafka/blob/f3c9666caa0ef1577402a5c16d57664bf06f51ef/flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java#L194) for the mentioned issue, please take a look. 
Test passes with this fixed code and on the latest main branch, it timeouts due to indefinite blocking.

**Comment by morazow** [comment]
Thanks @dongwoo6kim, Looks good!

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [inline flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:130]: This is not really clear to me. Are you discussing `if (consumerPosition > stoppingOffset) {`? 

Is the stop offset inclusive or exclusive? It looks like exclusive both in the old and the new version. But that's a bit unintuitive if I would like to read all data currently present in the Kafka topic and this approach discards it.

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [inline flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:131]: Afaik this is equivalent to the old `if (lastRecord.offset() >= stoppingOffset - 1) {`. This code gave the offset of the last method and your new version is using the offset of the next record.

Is this what you wanted? I'm not sure if this is fixing anything then except for the first fix above.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:375]: Why are you using Timeout here? We usually discourage the usage as any infra hiccup will result in test failure. Instead we rely on a global timeout where small infra hiccups are absorbed.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:383]: Why is this necessary?

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:403]: Is this failing with the old code?

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java:273]: Instead of code replication, how about patching `produceToKafka` to create transactions if transaction.id is set?

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:235]: Is it the deletion marker that would cause a hang?

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java
Comment (AHeise) [inline flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:238]: Again please remove all custom timeout logic.

**Review by AHeise — COMMENTED** [review]
@dongwoo6kim thank you very much for your contribution. I have a couple of remarks.

If you don't have time to fix it, I can also take over because we also hitting that on production.

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:131]: As you mentioned, `consumerPosition >= stoppingOffset` is generally equivalent to `lastRecord.offset() >= stoppingOffset - 1`, but there are some edge cases where the new condition works better.

### Edge case
#### Producer setup
Let’s say we are producing transactional messages, and the `lastRecord's offset` is `100`. 
After producing the lastRecord, we commit the transaction, which writes a `marker message` at offset `101`. 
Now, the `nextRecord's offset` becomes `102` (one more than the last message including marker message).

#### Consumer setup
When the flink consumer is reading with `scan.bounded.mode` set to `latest-offset`, the `stoppingOffset` is set to the `nextRecord's offset`, which in this case is `102`.

#### Comparison
In the old condition `(lastRecord.offset() >= stoppingOffset - 1)`, it would compare `100 >= 102 - 1`, which would never be met, leading to indefinite blocking.
However, with the new condition `(consumerPosition >= stoppingOffset)`, it compares `102 >= 102`, which is immediately satisfied. 
This happens because consumer.position() skips the marker message. 
Using lastRecord.offset() in the stopping condition is insufficient in transactional case (when a marker message follows the last record).


| Condition       | As-Is (Old Condition)                             | To-Be (New Condition)                             |
|-----------------|---------------------------------------------------|---------------------------------------------------|
| Expression      | `lastRecord.offset() >= stoppingOffset - 1`        | `consumerPosition >= stoppingOffset`              |
| Example Comparison | Compares `100 >= 102 - 1`, which never meets     | Compares `102 >= 102`, which is immediately satisfied |
| Reason          | The condition fails because it checks the last data record's offset, not accounting for the marker message | The condition succeeds because `consumer.position()` skips the marker message and correctly evaluates the stopping condition |
| Issue           | Leads to indefinite polling                       | Stops polling immediately

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:375]: I used the Timeout to prevent infinite blocking and fail the test if it enters an infinite loop. 
Since I'm not familiar with the project's global timeout — could you elaborate more about it? 
Where is it configured? 
If global timeout is set then it should be enough but when I tested this method on the current main branch without the Timeout, it didn't fail but just hung.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:383]: This is preprocessing step before using [executeAndVerify](https://github.com/apache/flink-connector-kafka/blob/7929b16dcfe648da30b6cc9755f63de2ed3d5319/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java#L482) method to verify test result. 

This method expects records from `partition P` should be an integer sequence from `P` to `NUM_RECORDS_PER_PARTITION`. So I deleted records where the value is less than the partition number.

Similar approach to [here](https://github.com/apache/flink-connector-kafka/blob/7929b16dcfe648da30b6cc9755f63de2ed3d5319/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java#L348)

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:403]: Yes this test fails with the current main branch due to timeout.

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java:235]: Yes by deleting message in this step we can cause hanging. 
This test code is for issue [here](https://issues.apache.org/jira/browse/FLINK-35565)

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaTestBase.java]: I agree, I've applied your idea. Thanks for the input

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:130]: It is inclusive `(consumerPosition >= stoppingOffset)`. I can't understand about discarding issue, could you elaborate more?

**Comment by dongwoo6kim** [comment]
@AHeise thanks for the feedback! 
I've addressed your comments and applied the suggested changes. When you have a moment, please take a look. Thanks!

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:384]: Could you change the comment to reflect how the data looks afterwards? I'm assuming this will preserve one record per partition?
I'm also curios why we need a specific data layout for this test to work? Or rephrased: what would happen if we retain all records originally generated? Wouldn't the test still assert similar things?

**Review by AHeise — COMMENTED** [review]
Thank you very much for the swift response and action. Changes look good to me. I will approve and merge once CI passes.

My only concern is that I don't fully understand why we need to modify the data in the test case. Shouldn't a transaction marker always be at the end? What did I miss?

Also if possible could you squash your commits and extend the commit message to include some of the information of the PR description? Basically state the problem and the briefly summarize the solution. Your commit message and PR title is already very descriptive, so it's not necessary for merging and it would just be a nice-to-have.

**Comment by AHeise** [comment]
Seems like the deleted arch unit rule was still needed. What was your intent when you deleted it?

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java:376]: The global timeout is applied during CI only. So after X minutes, the CI build will fail. If you test it locally, you have to abort manually (or you use the scripts to run the build as CI does).

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:127]: (This was just remark for other viewers of this PR).

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:130]: I think I understand now: 
* If I use stopping offset `endOffset` and it picks let's say 100, it would abort reading at 100 and sets the stopping offset of the partition to 100.
* The record 100 has been read and is part of the batch. It will not be filtered out.
* So it indeed appears as the last record of the batch.

File: flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java
Comment (AHeise) [reply flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/source/reader/KafkaPartitionSplitReader.java:131]: Thank you very much for the detailed explanation.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java]: Got it, thanks for the explanation. I've deleted all custom timeout logic in test codes

File: flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/streaming/connectors/kafka/table/KafkaTableITCase.java]: I've deleted timeout logic thanks for the feedback

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (dongwoo6kim) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java]: I've made two changes. 
1. Replaced `records.removeIf(record -> record.partition() > record.value())` with `KafkaSourceTestEnv.setupEarliestOffsets(transactionalTopic)`, as it serves a similar purpose.
2. Added a comment explaining how the data looks after the setup.

> I'm assuming this will preserve one record per partition?

After data modification, each partition p will contain records from p to NUM_RECORDS_PER_PARTITION (which is 10). For example, partition 1 has records 1 to 10, and partition 5 has records 5 to 10.
 
> what would happen if we retain all records originally generated

The main purpose of this data modification setup is to reuse `executeAndVerify` method. 
When you look at [here](https://github.com/apache/flink-connector-kafka/blob/7929b16dcfe648da30b6cc9755f63de2ed3d5319/flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java#L505) `executeAndVerify` method expects the input data to be modified like this way. 

I intended to reuse existing test util functions and follow the test code convention but if you think it is causing unnecessary confusion I can change the test code to have custom assertion logic.

**Comment by dongwoo6kim** [comment]
> Seems like the deleted arch unit rule was still needed. What was your intent when you deleted it?

It was automatically deleted after running `mvn clean verify` locally. I manually rolled back the archunit changes.

**Comment by dongwoo6kim** [comment]
@AHeise Thanks for the feedback. I've left some comments and made updates. Please have a look.

File: flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java
Comment (AHeise) [reply flink-connector-kafka/src/test/java/org/apache/flink/connector/kafka/source/KafkaSourceITCase.java]: If it's needed for verification, then all good. We should probably revise the setup then at a later point.

**Comment by AHeise** [comment]
Changes all LGTM. I'm running CI and merge when it's green. 
Thank you very much for this much needed fix!

**Pull request merged**
